consensus/istanbul/backend/announce.md

```diff
@@ -0,0 +1,140 @@
+# Celo Announce Protocol
+
+While here referred as a protocol by itself, Announce is technically a subset of the Istanbul protocol, on the [RLPx] transport, that facilitates Validators to privately discover the eNodeURL of other Validators in the network, using the underlying p2p eth network. This is used by Istanbul to have direct connections between Validators, speeding up the consensus phase of block generation.
+
+It is important to note that a validator's `eNodeURL` is never shared among third parties, but directly published to specific peers. Therefore it is allowed for a validator to publish different external facing `eNodeURL` values to different validator peers.
+
+## Basic Operation
+
+When a validator is close to being elected ([NearlyElectedValidator]), it starts periodically sending [queryEnodeMsg] messages through the p2p network. These message are regossipped by nodes, and validators reply with a direct message to the originator with an [eNodeCertificateMsg], holding their `eNodeURL`. The initial [queryEnodeMsg] contained the origin validator's `eNodeURL` encrypted with the destination public key, therefore after the direct reply, both validators are aware of each others' `eNodeURL`.
```

Comment:
```suggestion
When a validator is close to being elected ([NearlyElectedValidator]), it starts periodically sending [queryEnodeMsg]s through the p2p network. The [queryEnodeMsg] is essentially a mapping between a validator's Celo address and it's `eNodeURL` and is re-gossiped by all nodes, but the `eNodeURL` is hidden from the rest of the network by encrypting it. The `eNodeURL` is encrypted once for each other validator with that validator's public key (question does the query always contain the enode encyrpted for all other validators, or just some, and how is that decided?). Each encryption of the `eNodeURL` is marked with the corresponding validator's Celo address, so that validators receiving the message can tell which entry is intended for them. Upon receipt of this message a validator decrypts the encrypted `eNodeURL` and replies with a direct message to the originator with an [eNodeCertificateMsg], holding their own `eNodeURL`. So after the direct reply, both validators are aware of each others' `eNodeURL`
```

Reply:
`(question does the query always contain the enode encyrpted for all other validators, or just some, and how is that decided?)` -> that is explained in other parts of the spec. Should I link to that part at this stage? I was trying to be brief in this explanation.

I'm not sold on the sentence "The [queryEnodeMsg] is essentially a mapping between a validator's Celo address and its `eNodeURL`". Why do think it as a mapping?

Reply:
>`(question does the query always contain the enode encyrpted for all other validators, or just some, and how is that decided?)` -> that is explained in other parts of the spec. Should I link to that part at this stage? I was trying to be brief in this explanation.

I've seen it now, thanks but I think we could just say something like "The `eNodeURL` is encrypted once for each other validator that this validator needs to discover and it's encrypted with this validator's public key."

Its a bit of a mouthful, but I think it would help with the understanding, the previous description was a just a bit too light I thought to really understand the protocol. And trying to piece it together by looking at the individual behaviours is quite hard. Its nice to have a cohesive natural language description of what the protocol is doing.

Reply:
>I'm not sold on the sentence "The [queryEnodeMsg] is essentially a mapping between a validator's Celo address and its eNodeURL". Why do think it as a mapping?

As a receiver of the queryEnodeMsg the knowledge it provides to me is Validator address X has enode Y. Prior to receiving it the reciever simply knows that there is a validator with address X, the query enode message lets them complete the mapping. Does that make sense?